### ISSUE TITLE
How to implement web-compatible camera downscaling?

### ISSUE THREAD

**Issue by jan-ivar** [w3c/mediacapture-main#472]
The spec leaves it to UAs whether to offer downscaled resolutions and decimated frame-rates.

This leaves web developers with a problem. This:
```js
await navigator.mediaDevices.getUserMedia({video: {height: {exact: 360}}});
```

...means, depending on browser¹, either:
 1. Do you have a 360 native mode? If not, `OverConstrainedError` please, so I can do som'else, or
 2. Downscale the camera output to 360p, please.

Assuming both groups of users exist, how do we solve their respective problems in all browsers?

---
<sub>1. Firefox and Edge implement 1, while Chrome (even Canary w/new algorithm flag) implements 2. Safari didn't take constraints yet.</sub>

**Comment by jan-ivar** [comment]
Observation: The 2nd group has no need for `min`, `max`, `exact`, `ideal`, or `advanced`:
```js
await navigator.mediaDevices.getUserMedia({video: {height: 360}});
```
...is all they need. An inherently simple rescale-me-this API.

### Proposal
Give them that API, and force the 1st group to use `min`, `max`, `exact`, `ideal`, or `advanced`.

That is, have the presence of those keywords be a hint to ignore downscaled [settings dictionaries](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-settings-dictionary).

This would mean:
 * Only the group that needs them need learn the algorithm-related keywords.
 * `360` would downscale as needed (even in Firefox/Edge).
 * `{ideal: 360}` would find the closest native resolution (even in Chrome).
 * `{exact: 360}` would produce 360p native or `OverconstrainedError` (even in Chrome).
 * No downscaling in `advanced` (even in Chrome).
 * Should work for `frameRate` as well.

Note that this wouldn't otherwise change the meaning of plain values.

#### Examples
```js
await track.applyConstraints({height: 360}); // 360p downscaled if necessary
await track.applyConstraints({height: {ideal: 360}}); // closest native to 360p
await track.applyConstraints({height: {exact: 360}}); // native 360p or throw
await track.applyConstraints({height: {min: 320, max: 1920}}); // native in range or throw
await track.applyConstraints({height: {min: 320, max: 1920, ideal: 320}}); // lowest native
await track.applyConstraints({height: {max: 1920, ideal: 1920}}); // highest native
await track.applyConstraints({height: 9999}); // highest native available
```

Only the first is a change in Firefox/Edge. All but the first, middle and last a change in Chrome.

Win-win? The spec doesn't touch rescale behavior much, but we'd need a vendor agreement at least. It'd be confusing unless this worked the same in all browsers.

@guidou, @fluffy, @aboba, @jesup, @youennf WDYT?

**Comment by guidou** [comment]
What about having extra boolean constraints to control various forms of video processing.
We already have this in audio with constraints such as echoCancellation, noiseSupression and autoGainControl.
We could have additional frameRateAdjustment and rescaling constraints that work similarly for video.
This approach would have the advantage of consistency with the existing audio API, can be introduced progressively by browsers without altering the behavior of existing applications, and probably requires fewer spec changes.

**Comment by jan-ivar** [comment]
@guidou Interesting, but as a constraint, I think it fails, because while something like:
```js
await track.applyConstraints({height: 360, native: true});
```
...might give preference to native resolutions, the opposite:
```js
await track.applyConstraints({height: 360, native: false});
```
...would then give prefence specifically to non-native resolutions, which is always undesirable.

We'd also be bound by the [fitness-distance](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-fitness-distance) algorithm, so even with this `1` handicap, given that fake modes come with a perfect `0` score in `width`, `height`, `aspectRatio` and `frameRate`, we might not get the desired results in all cases.

Lastly, I'd like to reserve the UA's right to rescale in certain situations like https://github.com/w3c/mediacapture-main/issues/466#issuecomment-320375561, and `{native: {exact: true}}` might interfere with that.

As a boolean flag, I think it's redundant: `native: true` makes `min`, `max`, `exact`, `ideal` moot.

We didn't add `mandatory: true`, instead we overloaded the discovery keys `min`, `max` and `exact`.

Likewise, instead of `native: true` I think we should overload `min`, `max`, `exact`, and `ideal`.

**Comment by jan-ivar** [comment]
It also wouldn't resolve the current browser difference over
```js
await track.applyConstraints({width: 640, height: 360});
```
...which I was hoping to do. I was thinking something like:

*"The User Agent SHOULD omit downscaled or decimated settings dictionaries if constraints are specified using any of the keywords 'min', ‘max', 'exact' or ‘ideal'."*

**Comment by guidou** [comment]
Your example just shows that a "native" constraint where true means native and false means non-native would not be very useful.
If instead we have a "rescaling" or "rescalingAllowed" constraint, meaning that rescaling is allowed (but not required) when true would not have that problem. Having the constraint equal to exact false would have the effect of removing the candidate dictionaries with nonnative resolutions. 
Having the constraint as ideal false would result in native modes preferred over non-native modes, while allowing the possibility to switch to non-native in certain cases (e.g., concurrent access).
We can also make the constraint non-boolean if we need more than two values.

Changing the spec to redefine min/max/exact/ideal to mean native-only sounds like a bad idea to me. First, it introduces semantics that have no relation with what those words represent. It would seem counterintuitive to me that an application saying the ideal height is "638" means that 638 is forbidden.
Second, it is overly restrictive. It would mean that the only tracks that can be rescaled are unconstrained ones. I don't see why an application that specifies a range between 1-1000 should be forbidden from having a rescaled track in that range. Finally, what happens if min/max/exact is specified in some advanced set that is initially ignored, but becomes satisfiable because another track was closed? Does this mean that rescaling was initially allowed, but forbidden afterwards? This might require nontrivial changes to the spec to cover for various corner cases, or it would lead to many incompatible implementations. Another spec change that is needed is that naked values become a completely new category of value, distinct from ideal (in basic sets) and exact (in advanced sets).

The constraints-based solution I proposed seems a lot more straightforward to me. It allows the application to explicitly express intent. The only change required in the spec is to add a new entry to the list of supported constraints, not unlike the recent addition of processing constraints for audio. Also, it  opens the possibility to other processing techniques (e.g., some users have requested video mirroring, which is useful for many applications and currently requires a lot of boilerplate to implement, see issue #452).

**Comment by jan-ivar** [comment]
> I don't see why an application that specifies a range between 1-1000 should be forbidden from having a rescaled track in that range.

In Chrome, what's the difference between
```js
await track.applyConstraints({height: {min: 1, ideal: 638, max: 1000});
```
and
```js
await track.applyConstraints({height: 638});
```
?

**Comment by guidou** [comment]
That particular example Chrome would produce the same result for cameras (provided the camera always produces frames at the same resolution).
For screen capture, the first one will never exceed 1000 while the second might, depending on particular min/max values. In this paticular case, resolution will likely be adjusted to 640 for performance in both cases. With different values the results might differ.

Of course, for camera capture, I can easily produce examples combining width, height and aspect ratio where the selected resolution does not equal ideals, but where the results are different depending on the min/max values.  So, if your point is that ranges are in practice ignored in Chrome if you provide ideal is not valid in general.

**Comment by guidou** [comment]
Another difficulty I see with overloading min/max/exact to mean native mode is that resolution is a property determined by the combination of the width, height and aspectRatio constraints.
What happens if some of those constraints have naked values (indicating rescaling) and some have min/max (indicating native)?
In general, constraints control properties that are not always orthogonal, so changes in the range of one constraint can affect the selected value for another property. This is particularly true for width, height and aspectRatio, which together control resolution which is a single value (albeit 2D). Trying to control post-processing of the single output with a trick in (some of?) the constraints does not look like the right way to proceed, IMO. Especially since we have the precedent of doing similar things in audio using additional constraints.

**Comment by jan-ivar** [comment]
@guidou You mean, what would happen with:
```js
await navigator.mediaDevices.getUserMedia({video: {width: 640, height: {max: 360}}});
```
? - Going back to intent and the two user groups I mentioned, the rule would be any use of keywords implies an intent to discover discrete modes, so: native <= ???x360 or `OverconstrainedError`.

Mixing bare values with min/max/exact/ideal, and ideal values purposely outside of a min-max range are edge-cases with little insight into user intent IMHO.

I think the question to ask is: What can developers not express with this?

Remember the current spec says *nothing* about this, and I'm only advocating for a SHOULD. I still prefer UAs in charge of this decision, so they can handle concurrent access well.

A new constraint would undermine that, and do nothing to solve the implementation discrepancy developers are facing today, because constraints don't have defaults.

`width`, `height`, `aspectRatio` and `frameRate` are the only non-orthogonal constrainable properties, in my experience.

> screen capture in Chrome tends to adjust to standard aspect ratios if constraints allow.

Screen capture is a different spec. We can always discuss screenshare-specific constraints there.

**Comment by jan-ivar** [comment]
Since we don't upscale, we could make an exception for the `min` keyword, which may be useful to request a rescaled resolution of at least a certain size.

**Comment by guidou** [comment]
> I think the question to ask is: What can developers not express with this?

A constraint would allow using ideal to express a preference that is not a hard requirement.
For example, an application might prefer native modes, but be willing to accept non-native ones.
This could be easily expressed with constraints, but probably not with the keyword approach, unless you make another exception.

Constraints would give users fine control over multiple processing techniques.
For example, frameRate adjustment, resolution downscaling and even new effects that users have requested such as video mirroring would be easily controlled separately.

It is not clear to me that the keyword approach allows controlling frame-rate adjustment separately from rescaling, which an application might want to do. Native camera modes are normally triplets (width, height, frameRate), so if a camera has 640x480x60Hz as native mode, and the user requests max 30Hz, then it should throw OverconstrainedError since 640x480x30Hz is not a native mode, right?

> Remember the current spec says nothing about this, and I'm only advocating for a SHOULD. I still prefer UAs in charge of this decision, so they can handle concurrent access well. 
> A new constraint would undermine that, and do nothing to solve the implementation discrepancy developers are facing today, because constraints don't have defaults.

What would new processing constraints exactly undermine?
IMO, they would do a lot to solve implementation discrepancy in a straightforward and explicit way, for those who care about the discrepancy.
Users who don't care about the discrepancy are fine without using the constraints, even if browsers keep their current behavior as default. I suspect these are the majority since no one has filed bugs in Chrome about ideal not preferring native modes (for comparison, there were lots of popular bugs asking for support for ideal before Chrome 59). Do you have evidence that this discrepancy (in which all browsers correctly implement the spec) is negatively affecting users?

**Comment by jan-ivar** [comment]
@guidou You mentioned video mirroring again, but https://github.com/w3c/mediacapture-main/issues/452 was closed a while ago as out of scope. FWIW I agree with @burnburn that `getUserMedia` is not a post-processing API.

> there were lots of popular bugs asking for support for ideal before Chrome 59

What's the more likely reason for that popularity?
 1. People wanted Chrome to follow FF + Edge in implementing `ideal` as the spec intended, or
 2. They wanted min-max ranges around their downscaled resolutions, because ... screen-sharing?

I think there's a disconnect here.

> Do you have evidence that this discrepancy (in which all browsers correctly implement the spec) is negatively affecting users?

See above. - The main complaint we hear is developers expecting `getUserMedia` to produce exactly the dimensions they specify "because it works in Chrome". It's a completely different model whether one scales at the source or [the](https://stackoverflow.com/questions/36961228/how-to-keep-11-aspect-ratio-video-all-the-time-in-webrtc/36962250) [sink](https://blog.mozilla.org/webrtc/fiddle-week-downscale-video-peerconnection/).

> controlling frame-rate adjustment separately from rescaling,

I think the spec works well enough here, actually. Drivers can offer divide-by-two, divide-by-three etc. decimated frame-rates with few side effects, whereas apps wouldn't know which ones are good, and they're close enough for most app use. I think it would be a mistake to give continuous frame-rate control to the app.

These divide-by rates can be added by the UA as well if the driver doesn't offer them. So "native" doesn't have to mean 100% native, just discrete (good) choices rather than a continuous value space devoid of hints.

This model should be enough to offer close-enough rescaling/decimating to satisfy most bit-needs, while at the same time avoiding sub-par choices, unless one implements a continuous set of settings dictionaries, drowning out the good choices, like Chrome has.

**Comment by guidou** [comment]
> FWIW, I agree with @burnburn that getUserMedia is not a post-processing API.

So we should remove echoCancellation, autoGainControl and noiseSupression constraints because they are post-processing?

> What's the more likely reason for that popularity?
>
> 1. People wanted Chrome to follow FF + Edge in implementing ideal as the spec intended, or
> 2.  They wanted min-max ranges around their downscaled resolutions, because ... screen-sharing?

I can only read what people write in their bugs and what's written in the spec, not what's in other people's minds. Most bugs filed after the spec-compliant processing of video constraints shipped in Chrome 59 were about the behavior of screen sharing changing due to different default behavior with the new algorithm. Those bugs have been fixed since (while still conforming to the spec)  and I have not seen any other significant bugs about video-constraint processing.
More specifically, I have not seen evidence to suggest that the two reasons you present were what made the bugs popular, but, like I said, I can't read people's minds.

> The main complaint we hear is developers expecting getUserMedia to produce exactly the dimensions they specify "because it works in Chrome". It's a completely different model whether one scales at the source or the sink.

That sounds to me more like those users are OK with Chrome's implementation of the spec and not that they want Chrome to change its current implementation to match Firefox and Edge in this regard. Otherwise I think they would be filing Chrome bugs.
Also, I have seen Chrome users asking for video mirroring, which sounds at least reasonable to me since the spec has other post-processing constraints.

> I think the spec works well enough here, actually. Drivers can offer divide-by-two, divide-by-three etc. decimated frame-rates with few side effects, whereas apps wouldn't know which ones are good, and they're close enough for most app use. I think it would be a mistake to give continuous frame-rate control to the app.

Chrome already gives apps continuous frame-rate control and I have not seen users complain about it yet. I did see Chrome bugs filed when frame-rate control was not working correctly.

> These divide-by rates can be added by the UA as well if the driver doesn't offer them. So "native" doesn't have to mean 100% native, just discrete (good) choices rather than a continuous value space devoid of hints.

I agree that it's a good idea to allow applications to express that they prefer or require "native" values, even if no Chrome users have filed bugs asking for that feature. Our disagreement is about how to introduce that concept in the spec.

My opinion is that additional constrainable properties are a better fit for this than overloading the meaning of max/exact/ideal.

The constraints algorithm (as written in the spec) basically consists in:
1. Start with a set of valid candidates (each having a specific value for each constrainable property)
2. Discard the ones that do not satisfy the constraints specified by the app using min/max/exact.
3. Use ideal as the primary criterion to choose among the remaining candidates (ties among those that equally satisfy ideal are solved in an implementation specific manner). 

The nice thing about this approach is that it is generic enough that it applies equally to audio, video and potentially many other track types. Similarly, it is generic enough to allow many different types of constrainable properties (including post-processing ones, even if you think getUserMedia is not a post-processing API) without requiring special cases or exceptions for any of them. New filtering criteria or new types of candidates can be introduced with the addition of new constrainable properties with relatively low impact on the spec, existing applications and, possibly, implementations. 
My argument is that what we want to do in practice is add a new filtering type to Chrome and new candidate types (and the corresponding filtering) to Firefox and Edge. I think this is a good match for new constrainable properties.

My arguments against the overloading of min/max/exact/ideal are:
1. It adds nonintuitive meaning to those keywords, which may create user confusion.

2. It does not allow constrained non-native tracks. To overcome the first part you have to start adding exceptions (see 5).

3. It does not allow expressing a preference (without hard requirement) for native tracks. Wrt the second part, I guess you can change the definition to mean that min/max/exact means "prefer" but no hard requirement, in which case you need to make the spec even more complex by adding the concept of concurrency and explaining what is permitted under concurrency and what not.

4. It breaks existing applications. For example, many existing applications that use {echoCancellation: {exact: true}} or {echoCancellation: {ideal: true}} expect software echo cancellation (i.e., post-processing). If exact/ideal are redefined to also mean "native" or "no post-processing", then these applications would break unless we start adding exceptions for certain properties and/or certain types of tracks, making the algorithm not generic anymore. 

5. Once we start adding these specific exceptions we run the risk of creating more problems down the road. For example, it did not take long for you to propose an exception for min because you realized that users might want to constraint the minimum value of width or height of a rescaled resolution. What about constraining the maximum aspect ratio? What if we want to use audio resampling to implement the sampleRate constraint and want to support both downsampling and upsampling? Does that mean that we also need to introduce an exception for max and exact? Does that mean that the only overloaded keyword should be ideal, or that the exceptions would be different depending on the constraint and track type?
Honestly, I think this has the potential of making discrepancies of browser behavior worse due to bugs and different interpretations of the exceptions, especially when it comes to correlated properties (different browsers might interpret different sets of properties as correlated) and different treatment of concurrent and non-concurrent scenarios.

So far, the arguments you have brought  against introducing new constrainable properties are:
1. The original intention of the authors of the spec was to mandate Firefox-like behavior (native-mode disovery). My answer to that is that this is irrelevant now because this intention is not expressed in any way whatsoever in the part of the spec that defines the constrainable pattern. The question now is how to introduce this intention in the spec with minimal disruption to existing applications, implementations and the spec itself, given there is already a considerable investment in the spec as currently written by both application developers and implementation writers. 
2. The proposed constraints introduce post-processing and getUserMedia is not a post-processing spec. My answer is that, as written, getUserMedia is a post-processing spec. There is language in the spec that says "User Agent processing, such as downsampling, may be used to ensure that all tracks have appropriate media." Also, there are constrainable properties that explicitly refer to post-processing techniques, such as echoCancellation, autoGainControl and noiseSuppression.

Your arguments in favor of overloading min/max/exact/ideal:
1. It matches the original intention of the spec authors. Like I said, I find this irrelevant since that intention is neither explicitly nor implicitly expressed in the spec. I would argue that the genericness of the spec wrt constraints suggests to me that another intention of the authors was to keep it generic and free from exceptions and concepts specific to particular track types or properties. This opinion is based on what's actually written in the spec.
2. The proposed changes are a SHOULD. I think this is a disadvantage. This would mean that some browsers might opt to fully implement this SHOULD, others might implement it in some circumstances (e.g., only when there is no concurrent access), while others might not implement it at all. This could result in a even more divergent browser behavior together with a more complex spec, potentially making the problem worse.

My conclusion is that the case for using new constrainable properties to control rescaling is stronger than giving extra meaning to min/max/exact/ideal because the former has much lower compatibility and interoperability risks. Moreover, I think that leaving the spec unmodified and letting the browser choose how to do rescaling is also better than redefining min/max/exact/ideal. Nothing in the current spec prevents Firefox from, for example, maintaining its current behavior in the non-concurrent case and resorting to rescaling only when concurrency and constraints prevent native mode for all tracks.

I apologize if I misunderstood or forgot to include some of your arguments, so please feel free to correct me and add new thoughts so that we can advance the discussion.

**Comment by jan-ivar** [comment]
Let's contain the scope to keep this productive. To clarify my [problem statement](https://github.com/w3c/mediacapture-main/issues/472#issue-248089527), my intent here is only to try to bridge the gap between *existing browser features*. For new problems, open new issues:

 * Video mirroring is https://github.com/w3c/mediacapture-main/issues/452. Let's agree to disagree in that issue, not here.
 * Screen-sharing has its own [spec](https://w3c.github.io/mediacapture-screen-share/#constraints) which doesn't even allow constraints for initial selection, making most of your arguments there moot. Users read `track.getSettings()` modify them and set them with back with `track.applyConstraints()`. No need for `min`, `max`, `exact`, `ideal` or `advanced`. Disagree? Please open an issue in that spec.
 * Sky isn't falling. https://github.com/w3c/mediacapture-main/issues/473 adds a SHOULD around some existing browser behavior. *Reduces* entropy.
 * There's no gap in microphone constraints. No change there.
 * New features from new processing constraints are out of scope. File a separate issue please.
 * This thread is getting long. Let's try to be brief. People can scroll to see earlier arguments.

> I agree that it's a good idea to allow applications to express that they prefer or require "native" values, even if no Chrome users have filed bugs asking for that feature. 

That problem seems unique to Chrome, because of its choice to not fully utilize the full expressiveness of the constraints language.

> Our disagreement is about how to introduce that concept in the spec.

That concept is already in the spec, as evidenced by Firefox and Edge users not having this problem.

**Comment by jan-ivar** [comment]
> What about constraining the maximum aspect ratio?

If you're asking if FF would consider rescaled modes or not, I'd say we ["can use any information available"](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-selectsettings) to make that decision. We don't need agreement for that. It'd be nice if our behaviors aligned better however. That's why I opened this issue.

For instance, FF could appease the we-want-what-we-want group with rescaling in this case:
```js
await navigator.mediaDevices.getUserMedia({video: 640, height: 360}); // downscales 640x480
```
while still leaving everything else as-is:
```js
await navigator.mediaDevices.getUserMedia({video: 640, height: {exact: 360}}); // error
```

It'd be internally inconsistent, and thus confusing, but as much as we all like to admire internal consistency, we tend to forget that cross-browser inconsistency is often equally confusing. I'm exploring sacrificing some internal consistency to achieve cross-browser consistency.

To turn this around, assuming no interest in a new constraint, what might you be interested in pursuing to narrow the current browser gap in output from the constraints algorithm?

**Comment by guidou** [comment]
> That  problem seems unique to Chrome, because of its choice to not fully utilize the full expressiveness of the constraints language.

I don't think Chrome has a problem with this. No one has filed Chrome bugs against this behavior. 

> That concept is already in the spec, as evidenced by Firefox and Edge users not having this problem.

No, the concept of native modes or their discovery is not in the spec. Chrome supports rescaling and its constraints algorithm returns rescaled resolutions over native ones when they result in lower fitness values as defined by the spec.
FF and Edge only support native resolutions and they return native resolutions. There is nothing in the spec to suggest that native modes should be preferred over rescaled ones when the rescaled ones have a better fitness. The fact that you we are discussing possible spec changes in order to support both rescaling and native modes over rescaled modes just confirms this.

> To turn this around, assuming no interest in a new constraint, what might you be interested in pursuing to narrow the current browser gap in output from the constraints algorithm?

Currently the main priority for Chrome is to be spec compliant. There are many parts of the spec that Chrome does not implement and the priority is to focus on those. The browser gap is a lot greater in those areas and that is therefore the priority for Chrome.
With respect to the output of the constraints algorithm, it is spec compliant and users have not filed bugs so far complaining about it. Thus, changes in this area have low priority.
In particular, we are not willing to pursue the proposal of changing the semantics of the constraint keywords due to the reasons already explained. We are willing to consider other proposals that do not change the meaning of the keywords, even if they are not based on new constraints. 
An option that comes to mind is changing the fitness function so that it penalizes nonnative candidates more than native ones, but with a finite penalty so that they still satisfy the constraint and without redefining keywords.
Something like 
  fitness = (requires_processing ? 1000.0 : 0.0) + |actual -ideal|/max(|actual|,|ideal|)

This would actually introduce the native-mode discovery that the authors originally intended and it would still allow users to control rescaling. Applications that require a specific (potentially rescaled) resolution would use exact on the basic set. Applications that prefer but do not require rescaled would use exact on an advanced constraint.
I still have to analyze the full consequences of such a change, especially but, in principle, it seems a lot more reasonable than redefining keywords and naked values.

**Comment by jan-ivar** [comment]
> Applications that require a specific (potentially rescaled) resolution would use exact

@guidou That might work. You don't need to change the fitness algorithm even. The spec says *"The UA SHOULD use the one with the smallest fitness distance"*, meaning you don't have to, and *"the UA can use any information available to it. Examples ... may be ... whether using a settings dictionary will cause the device driver to apply resampling."*

My spec suggestion was mostly to rally convergence. As we've seen, the spec is fairly flexible here.

**Comment by guidou** [comment]
@jan-ivar If we are going agree on this approach in order to facilitate convergence, I think we should express it in the spec somehow as reference for developers and implementors. As currently written,  implementing the current SHOULD about minimizing fitness with the current fitness function means that implementations SHOULD do what Chrome currently does. Of course, they don't have to, but it would be good to have a SHOULD that matches the intent of the spec.

**Comment by guidou** [comment]
@jan-ivar Another, change I would be willing to consider is to make naked values in the basic set mean exact instead of ideal, just as in the advanced sets. This might be more intuitive to users who are less familiar with the spec and expect {width: 638} to mean exactly 638 and not native closest to 638.

**Comment by stefhak** [comment]
FWIW, my memory says alternative 2. in https://github.com/w3c/mediacapture-main/issues/472#issue-248089527 is what we had in mind back when this was originally specced. @burnburn would be good to get your input here!

Also note @fluffy's comment regarding "native" resolution in https://github.com/w3c/mediacapture-main/issues/466#issuecomment-321050942

**Comment by jan-ivar** [comment]
@guido I imagine bare values as exact would meet resistance since that was shot down years ago. We don't want to fail in simple cases, like:
```js
getUserMedia({video: {width: 2880}}) // highest available, not error
```

**Comment by burnburn** [comment]
Yeah, we argued for a long time on this and decided that bare values meant exact.

**Comment by alvestrand** [comment]
I see that there's been a long discussion here already. I'll review the discussion before making further comments, but I will make comments.

**Comment by jan-ivar** [comment]
### 2nd Proposal

The [current idea we're working on](https://bugzilla.mozilla.org/show_bug.cgi?id=1286945#c10)  in Firefox is to crop and downscale only when the app forces us to using `{max}` or `{exact}` where today we'd throw `OverconstrainedError`. This seems like a safe and intuitive interpretation that's equally spec-compliant, yet doesn't disrupt discovery of actual modes.

```js
await track.applyConstraints({width: 640, height: 360}); //640x480 not downscaled/cropped 
await track.applyConstraints({width: {exact: 640}, height: {exact: 360}); //640x360 cropped
```
(Of course when 640x360 is available as a driver mode, both lines above would produce 640x360).

This is based on "ideal has gravity" being a means of discovery, not a black hole.

**Comment by jan-ivar** [comment]
I agree w/@guidou [here](https://github.com/w3c/mediacapture-main/issues/472#issuecomment-321167341) that *"it would be good to have a SHOULD that matches the intent of the spec."*

This seems like a reasonable ask. We should verify the WG's intent here.

**Comment by alvestrand** [comment]
So that I can find it again while discussing:

The kind-of-similar CSS property is called "object-fit", and the variants "none" and "cover" are the two that match the two modes that are being proposed in #502, I think.

https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit

**Comment by alvestrand** [comment]
VI decision: merge #502 (add the constraint). No default for browsers is mandated.

**Comment by alvestrand** [comment]
Solved by #502
